src/Components/Components/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IAsyncDisposable
+    {
+        private int _itemsAbove;
+
+        private int _itemsVisible;
+
+        private ElementReference _topSpacer;
+
+        private ElementReference _bottomSpacer;
+
+        private IVirtualizationHelper? _virtualizationHelper;
+
+        [Inject]
+        private IVirtualizationService VirtualizationService { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the total number of items in the collection.
+        /// </summary>
+        protected abstract int ItemCount { get; }
+
+        /// <summary>
+        /// Gets the size (height) of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="RenderFragment"/>s representing a range of items.
+        /// </summary>
+        /// <param name="start">The start index of the range of items.</param>
+        /// <param name="count">The number of items in the range.</param>
+        /// <returns>The <see cref="RenderFragment"/>s representing the given range.</returns>
+        protected abstract IEnumerable<RenderFragment> GetItems(int start, int count);
```

Comment:
TBH, having the contract be defined in terms of `IEnumerable<RenderFragment>` instead of `IEnumerable<TItem>` is a bit strange and makes it harder to have features like "the child content template can be passed as *either* `ChildContent` or `Item`".

I see this has been done to simplify fitting `InfiniteScroll` into the system (so it can append the `InfiniteScrollFooter`) but I still wonder if we could do without that. For example if the infinite scroll footer was achieved by the developer customizing a `SpacerAfter` parameter (type `RenderFragment`) as per the alternative design suggestion, then we wouldn't need to append it to the list and hence could more easily have the contract defined in terms of `IEnumerable<TItem>`.